### PR TITLE
Feat/adapt bulker to safe

### PR DESCRIPTION
**Pull request by oumar-fall**
### Pull-Request Checklist

- [ ] Code is up-to-date with the `main` branch
- [ ] This pull request links relevant issues as `Fixes #0000`
- [ ] There are new or updated unit tests validating the change
- [ ] Documentation has been updated to reflect this change
- [ ] The new commits follow conventions outlined in the [conventional commit spec](https://www.conventionalcommits.org/en/v1.0.0/)

### CODE REVIEW

**Review by julien-devatom — COMMENTED** [review]
This is ok to me, but Ive an architectural improvement proposal:

- define an abstract BastTxBuilder class, defining protected functions as abstract, and build `Safe.TxHandler` & `Bulker.TxHandler` on top

**Comment by ghivert** [comment]
> This is ok to me, but Ive an architectural improvement proposal:
> 
> * define an abstract BastTxBuilder class, defining protected functions as abstract, and build `Safe.TxHandler` & `Bulker.TxHandler` on top

I'd agree, but if we manage more than two TxHandler. Right now I think the form is really more simple, and avoid to fetch a long inheritance chain. To me, it just adds more complexity for no concrete benefits. It will still be time to define an abstract class later, with a (maybe) third TxHandler. Moreover, the Interfaces are there exactly for that, and we already defined a `IBatchTxHandler`.

**Review by ghivert — APPROVED** [review]

**Comment by oumar-fall** [comment]
> > This is ok to me, but Ive an architectural improvement proposal:
> > 
> > * define an abstract BastTxBuilder class, defining protected functions as abstract, and build `Safe.TxHandler` & `Bulker.TxHandler` on top
> 
> I'd agree, but if we manage more than two TxHandler. Right now I think the form is really more simple, and avoid to fetch a long inheritance chain. To me, it just adds more complexity for no concrete benefits. It will still be time to define an abstract class later, with a (maybe) third TxHandler. Moreover, the Interfaces are there exactly for that, and we already defined a `IBatchTxHandler`.

Too late haha

I had mixed feelings towards these two positions. I decided to change it because there's no emergency and I thought it would be a good exercise in term of code design for blue.

I think we should use these opportunities to try new things that are obviously cleaner

**Review by ghivert — APPROVED** [review]

**Comment by ghivert** [comment]
> Too late haha
> 
> I had mixed feelings towards these two positions. I decided to change it because there's no emergency and I thought it would be a good exercise in term of code design for blue.
> 
> I think we should use these opportunities to try new things that are obviously cleaner

I get your point. I didn't disagree with the choice to isolate the BaseBulker. I think we should just pay attention to not over-engineer, and to avoid cognitive overload (the more layer you have, the more complexity you have too), and that we could have kept things as they were.
The final form is great, and I think the change is good. It was just not mandatory for me, but maybe as an FP guy, having layer of inheritance is complicated for my mind. 😅

File: src/txHandler/batch/Safe.TxHandler.ts [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+                to: CONTRACT_ADDRESSES.wsteth,
+                value: "0",
+                data: wstEthInterface.encodeFunctionData("unwrap", [
+                  operation.amount,
+                ]),
+              };
+            }
+          }
+          default: {
+            throw Error(`${operation.type} not implemented`);
```
Comment (julien-devatom) [inline src/txHandler/batch/Safe.TxHandler.ts:212]:
```suggestion
            throw Error(`${operation.type} not implemented for a safe`);
```

File: src/txHandler/batch/Safe.TxHandler.ts [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+
+    if (getAddress(underlyingAddress) === CONTRACT_ADDRESSES.wsteth) {
+      const wstethMissing = maxBN(
+        amount.sub(userMarketsData[underlyingAddress]!.walletBalance),
+        constants.Zero
+      );
+
+      if (wstethMissing.gt(0)) {
+        // we need to wrap some stETH
+        // To be sure that  , we add a buffer to the amount wrapped
```
Comment (julien-devatom) [inline src/txHandler/batch/Safe.TxHandler.ts:355]:
```suggestion
        // To be sure that we wrap enough tokens for the tx, we add a buffer to the amount wrapped
```

**Review by julien-devatom — COMMENTED** [review]
Test is failing due to a rounding. Ive some questions but nothing critical, congrats 🔥

**Review by julien-devatom — APPROVED** [review]

**Pull request merged**